File: Algorithms/Searching-Algorithm/Trie-Search/Trie_Search.py
# Program for Trie Insert and Search

# Trie node
class TrieNode:
	def __init__(self):
		self.children = [None]*26
		self.isEndOfWord = False

# Trie Class
class Trie:
	def __init__(self):
		self.root = self.getNode()

	def getNode(self):
		return TrieNode()

	def toIndex(self, ch):
		return ord(ch)-ord('a')

	# Function to Insert into Trie
	def insert(self, key):
		node = self.root
		length = len(key)

		for i in range(length):
			idx = self.toIndex(key[i])
			if not node.children[idx]:
				node.children[idx] = self.getNode()
			node = node.children[idx]
		node.isEndOfWord = True

	# Function for Trie Search
	def search(self, key):
		node = self.root
		length = len(key)

		for i in range(length):
			idx = self.toIndex(key[i])
			if not node.children[idx]:
				return False
			node = node.children[idx]
		return node!=None and node.isEndOfWord


# Main function 
# To test code, please edit keys list and print statements for search
def main():
	keys = ["the","apple","there","hello","world", "by","python"]
	output = ["Not present in trie", "Present in trie"]

	# Construct Trie
	trie = Trie()

	for key in keys:
		trie.insert(key)


	# Search in trie
	print("{} => {}".format("the", output[trie.search("the")]))
	print("{} => {}".format("hi", output[trie.search("hi")]))
	print("{} => {}".format("apple", output[trie.search("apple")]))
	print("{} => {}".format("man", output[trie.search("man")]))

if __name__ == '__main__':
	main()

## Sample Output ##
# the => Present in trie
# hi => Not present in trie
# apple => Present in trie
# man => Not present in trie

File: Algorithms/Searching-Algorithm/Trie-Search/Trie_Search.go
package main

import (
	"fmt"
	"strings"
)

// Trie structure
type Trie struct {
	root *trieNode
}

type trieNode struct {
	children [26]*trieNode
	wordEnd  bool
}

// Constructor initialize the Trie data structure
func Constructor() Trie {
	return Trie{
		root: &trieNode{},
	}
}

// Insert a word into the trie
func (this *Trie) Insert(word string) {
	word = strings.ToLower(word)
	current := this.root
	for i := 0; i < len(word); i++ {
		index := word[i] - 'a'
		if current.children[index] == nil {
			current.children[index] = &trieNode{}
		}
		current = current.children[index]
	}
	current.wordEnd = true
}

// Search returns if the word is in the trie
func (this *Trie) Search(word string) bool {
	word = strings.ToLower(word)
	current := this.root
	for i := 0; i < len(word); i++ {
		index := word[i] - 'a'
		if current.children[index] == nil {
			return false
		}
		current = current.children[index]
	}
	if current.wordEnd {
		return true
	}
	return false
}

func main() {
	insertWords := []string{"hello", "world", "golang", "trie", "tree"}
	searchWords := []string{
		"world",  /* true */
		"Golang", /* true */
		"word",   /* false */
	}

	trie := Constructor()

	for _, node := range insertWords {
		trie.Insert(node)
	}

	for _, node := range searchWords {
		found := trie.Search(node)
		if found {
			fmt.Println(node, ": Found in the trie")
		} else {
			fmt.Println(node, ": Not found in the trie")
		}
	}
}
